### ISSUE TITLE
Combinators to recover from defects

### ISSUE THREAD

**Issue by gabro** [zio/zio#3063]
I have an application which is a client-facing HTTP API. This application communicates with an external service which may occasionally be unreachable.

Whenever this happens and kill the fiber with `IO.die(ServiceFooNotAvailable`: I don't want this error to appear in my "domain errors" since my business logic can't do anything about it.

I do need however to translate this error to something intelligible for my HTTP clients, so at the "end of the world" I want to transform this defect into a failure and deal with it.

The best way I could find (and discussing on Discord it seems I didn't miss more elegant ways to do it) is something like:

```scala
def myMethod: IO[String, Int] = IO.die(ServiceFooNotAvailable)

// ... later ...

myMethod.catchSomeCause {
  case Cause.Traced(Cause.Die(_: ServiceFooNotAvailable), _) => IO.succeed(42)
}
```

or (probably better)

```scala
myMethod.catchSomeCause {
  case c if c.dieOption.exists(_.isInstanceOf[ServiceFooNotAvailable]) => IO.succeed(42)
}
```

Long story short, it would be nicer to have a direct combinator to recover from a dead fiber. Something along the lines of

```scala
def catchSomeDieCause[R1 <: R, E1 >: E, A1 >: A](pf: PartialFunction[Throwable, ZIO[R1, E1, A1]]): ZIO[R1, E1, A1]
```

**Comment by adamgfraser** [comment]
I will take this. One question is what the semantics should be if there is more than one cause. It seems most intuitive to me that we should refail if there are additional causes beyond the one being "caught" but that is different from your second implementation.

**Comment by iravid** [comment]
I think it's fair to use `find` semantics; that is, `catchSomeDie` / `catchAllDie` handle the defect if the function matches just one of the causes.

I think that is what `catchSome` does for `Both(Fail, Fail)`.

**Comment by adamgfraser** [comment]
Okay I will take a look. Just checking the first one seems very dubious.

**Comment by gabro** [comment]
I agree that matching on any of the causes seems reasonable from a user perspective

**Comment by jdegoes** [comment]
I'd suggest this family of combinators, if it existed, would be the inverse to `refineOrDie` family.

`refineOrDie` family takes part of the typed error channel and converts it into a non-recoverable error.

You can imagine an `unrefine` family taking part of the non-recoverable error and converting it into the typed error channel.

Symmetry between these could be a guiding design principle of the combinators.

**Comment by gabro** [comment]
I like the idea of symmetry! That said I think this family of combinators would be symmetrical do the "die" part of `refineOrDie`, since there's no "unrefinement" from `E` to `Throwable` going on (if anything, the opposite).

The symmetric with `die` made me think of a possible name:

```scala
def revive[R1 <: R, E >: E, A1 >: A](f: Throwable => ZIO[R1, E1, A1]): ZIO[R1, E1, A1]

def reviveSome[R1 <: R, E1 >: E, A1 >: A](pf: PartialFunction[Throwable, ZIO[R1, E1, A1]]): ZIO[R1, E1, A1]
```

what do you all think?
